jgrapht-core/src/main/java/org/jgrapht/graph/builder/GraphTypeBuilder.java

```diff
@@ -0,0 +1,341 @@
+ * 
+ * </blockquote>
+ * 
+ * Similarly one could get a weighted multigraph by using: <blockquote>
+ * 
+ * <pre>
+ * Graph&lt;Integer, DefaultWeightedEdge&gt; g = GraphTypeBuilder
+ *     .&lt;Integer, DefaultWeightedEdge&gt; undirected().allowingMultipleEdges(true)
+ *     .allowingSelfLoops(false).edgeClass(DefaultWeightedEdge.class).weighted(true).buildGraph();
+ * </pre>
+ * 
+ * </blockquote>
+ * 
+ * <p>
+ * The builder also provides the ability to construct a graph from another graph such as:
+ * <blockquote>
+ * 
+ * <pre>
+ * Graph&lt;Integer, DefaultWeightedEdge&gt; g1 = GraphTypeBuilder
+ *     .&lt;Integer, DefaultWeightedEdge&gt; undirected().allowingMultipleEdges(true)
+ *     .allowingSelfLoops(false).edgeClass(DefaultWeightedEdge.class).weighted(true).buildGraph();
+ * 
+ * Graph&lt;Integer, DefaultWeightedEdge&gt; g2 = GraphTypeBuilder.asGraph(g1).buildGraph();
+ * </pre>
+ * 
+ * </blockquote>
+ * 
+ * @param <V> the graph vertex type
+ * @param <E> the graph edge type
+ * 
+ * @author Dimitrios Michail
+ * 
+ * @see GraphType
+ * @see GraphBuilder
+ */
+public final class GraphTypeBuilder<V, E>
+{
+    private boolean undirected;
+    private boolean directed;
+    private boolean weighted;
+    private boolean allowingMultipleEdges;
+    private boolean allowingSelfLoops;
+    private EdgeFactory<V, E> edgeFactory;
+
+    private GraphTypeBuilder(boolean directed, boolean undirected)
+    {
+        this.directed = directed;
+        this.undirected = undirected;
+        this.weighted = false;
+        this.allowingMultipleEdges = false;
+        this.allowingSelfLoops = false;
+    }
+
+    /**
+     * Create a graph type builder for a directed graph.
+     * 
+     * @return the graph type builder
+     * @param <V> the graph vertex type
+     * @param <E> the graph edge type
+     */
+    public static <V, E> GraphTypeBuilder<V, E> directed()
+    {
+        return new GraphTypeBuilder<>(true, false);
+    }
+
+    /**
+     * Create a graph type builder for an undirected graph.
+     * 
+     * @return the graph type builder
+     * @param <V> the graph vertex type
+     * @param <E> the graph edge type
+     */
+    public static <V, E> GraphTypeBuilder<V, E> undirected()
+    {
+        return new GraphTypeBuilder<>(false, true);
+    }
+
+    /**
+     * Create a graph type builder for a mixed graph.
+     * 
+     * @return the graph type builder
+     * @param <V> the graph vertex type
+     * @param <E> the graph edge type
+     */
+    public static <V, E> GraphTypeBuilder<V, E> mixed()
+    {
+        return new GraphTypeBuilder<>(true, true);
+    }
+
+    /**
+     * Create a graph type builder which will create a graph with the same type as the one provided.
+     * 
+     * @param type the graph type
+     * @return the graph type builder
+     * @param <V> the graph vertex type
+     * @param <E> the graph edge type
+     */
+    public static <V, E> GraphTypeBuilder<V, E> asGraphType(GraphType type)
```

Comment:
The method naming convention "asX" usually pairs with a conversion to type X.  So I suggest "forGraphType" and "forGraph" instead.